### DISCUSSION TITLE
Dynamic Root List - Data Binding with DevExpress Blazor DataGrid

### DISCUSSION THREAD

**Discussion by Vickidh2000** [MarimerLLC/csla#4389] [Questions]
Dears,
I'm working on a CSLA Application and I would like to use a Dynamic Root List  as Business Object to bind with a DevExpress Blazor DataGrid. How this can be implemented ?  The  Dynamic Root List implement a kind of INotifyPropertyChanged  like it is in MVVM pattern with WPF or the implementation is different ? The instance of the Business Object should be link to the Data attribute of the DataGrid ?
 (Version  .net 8 & Blazor CSLA 8)

**Comment by rockfordlhotka** [comment]
The purpose of `DynamicListBase` is to retrieve a list of _editable root_ objects and bind them into some UI model (like a data grid) such that when the user leaves a row in the data grid, the root object bound to that row is immediately saved.

The _original_ behaviors were all built for Windows Forms and the built-in data grid control. That was the `DynamicBindingListBase` type.

The `BindingListBase` type has the same basic _concept_, but to my knowledge there are no data grid controls that follow the old data binding interface standard to make this work out of the box. As a result, it is always up to the app developer to write code to handle the three key events that are implemented by `BindingListBase`:

1. `BeginEdit` - this is invoked when the user enters a row in the data grid, and it indicates that the editable root object bound to that row is going to be edited by the user
2. `CancelEdit` - this is invoked when the user leaves the row in the data grid in such a way that they have indicated they don't want to save any of their changes, and that those changes should be rolled back (this is done using n-level undo)
3. `ApplyEdit`/`EndEdit` - this is invoked when the user leaves the row in the data grid in such a way that they have indicated they do want to save their changes; this triggers the immediate save of the editable root object bound to the row in the data grid

It is up to you to identify the events raised by the data grid control, handle those events in your code, and invoke the appropriate method as listed above to tell your dynamic list object to do the right thing.
